metadata.go

```diff
@@ -0,0 +1,221 @@
+package sarama
+
+import (
+	"sync"
+)
+
+type MetadataRefresh func(topics []string) error
+
+// currentRefresh makes sure sarama does not issue metadata requests
+// in parallel. If we need to refresh the metadata for a list of topics,
+// this struct will check if a refresh is already ongoing, and if so, it will
+// accumulate the list of topics to refresh in the next refresh.
+// When the current refresh is over, it will queue a new metadata refresh call
+// with the accumulated list of topics.
+type currentRefresh struct {
+	sync.Mutex
+	ongoing   bool
+	topicsMap map[string]struct{}
+	topics    []string
+	allTopics bool
+	chans     []chan error
+
+	// This is the function that gets called when to refresh the metadata.
+	// It is called with the list of all topics that need to be refreshed
+	// or with nil if all topics need to be refreshed.
+	refresh func(topics []string) error
+}
+
+func newCurrentRefresh(f func(topics []string) error) *currentRefresh {
+	mr := &currentRefresh{
+		topicsMap: make(map[string]struct{}),
+		topics:    make([]string, 0),
+		refresh:   f,
+		chans:     make([]chan error, 0),
+	}
+	return mr
+}
+
+// AddTopicsFrom adds topics from the next refresh to the current refresh.
+// You need to hold the lock to call this method.
+func (r *currentRefresh) AddTopicsFrom(next *nextRefresh) {
+	if next.allTopics {
+		r.allTopics = true
+		return
+	}
+	if len(next.topics) > 0 {
+		r.AddTopics(next.topics)
+		next.topics = next.topics[:0]
+	}
+}
+
+// nextRefresh holds the list of topics we will need
+// to refresh in the next refresh.
+// When a refresh is ongoing, calls to RefreshMetadata() are
+// accumulated in this struct, so that we can immediately issue another
+// refresh when the current refresh is over.
+type nextRefresh struct {
+	sync.Mutex
+	topics    []string
+	allTopics bool
+}
+
+// AddTopics adds topics to the refresh.
+// You need to hold the lock to call this method.
+func (r *currentRefresh) AddTopics(topics []string) {
+	if len(topics) == 0 {
+		r.allTopics = true
+		return
+	}
+	for _, topic := range topics {
+		if _, ok := r.topicsMap[topic]; ok {
+			continue
+		}
+		r.topicsMap[topic] = struct{}{}
+		r.topics = append(r.topics, topic)
+	}
+}
```

Comment:
I’m not sure I see what value there is in keeping two copies of the `topics`. One as a slice, and the other as a map. We don’t need to order them in the outgoing request, right?

And since this is the only possible place we could be assigning into the map, and we had to grab a mutex to even be, we can lazy initialize the map:
```suggestion
func (r *currentRefresh) addTopics(topics []string) {
	if len(topics) == 0 {
		r.allTopics = true
		return
	}
	if r.topics == nil {
		// We know we're going to be trying to assign at least one value into this map.
		// So we can just test ahead of each addTopics.
		r.topics = make(map[string]struct{})
	}
	for _, topic := range topics {
		r.topics[topic] = struct{}{} // faster to just assign directly rather than test first
	}
}
```

Reply:
I mean, there’s no reason why we cannot just use the map, and iterate over the map to make the slice to send?